### PR TITLE
Making memcached optional (for development and tests)

### PR DESCRIPTION
**Pull request by bartaz**
If `MEMCACHED_HOST` is not configured launchpad handler will just pass through memcached calls without caching anything.

This allows to run dev server without running memcached server and test without stubbing memcached.

### CODE REVIEW

File: src/server/helpers/memcached.js [hunk 1]
```diff
@@ -0,0 +1,30 @@
+import Memcached from 'memcached';
+
+import { conf } from '../helpers/config';
+
+let memcached = null;
+
+const getMemcachedStub = () => {
+  return {
+    get: (key, callback) => callback(),
+    set: (key, value, lifetime, callback) => callback()
```
Comment (cjwatson) [inline src/server/helpers/memcached.js]: I'd suggest making the default stub use a trivial in-memory cache object, the way that some tests do.  That would make it a more accurate stub and would simplify tests further.

File: src/server/helpers/memcached.js [hunk 2]
```diff
@@ -0,0 +1,30 @@
+import Memcached from 'memcached';
+
+import { conf } from '../helpers/config';
+
+let memcached = null;
+
+const getMemcachedStub = () => {
+  return {
+    get: (key, callback) => callback(),
```
Comment (bartaz) [inline src/server/helpers/memcached.js]: Just a simple 'pass-through' stub - nothing is cached, only callback is called (with undefined value).

File: src/server/helpers/memcached.js [hunk 3]
```diff
@@ -0,0 +1,30 @@
...
+    set: (key, value, lifetime, callback) => callback()
+  };
+};
+
+export const getMemcached = () => {
+  const host = conf.get('MEMCACHED_HOST') ? conf.get('MEMCACHED_HOST').split(',') : null;
```
Comment (bartaz) [inline src/server/helpers/memcached.js]: Should we require `MEMCACHED_HOST` to be defined in production environment?
Or is assuming that config will be complete is good enough?

File: test/routes/src/server/routes/t_launchpad.js [hunk 4]
```diff
@@ -473,25 +458,58 @@ describe('The Launchpad API endpoint', () => {
...
+          callback(undefined, true);
+        };
+        setMemcached(memcachedStub);
+
+        memcachedStub.cache[`url:${repositoryUrl}`] = snapUrl;
+      });
+
+      afterEach(() => {
+        setMemcached(null);
+      });
```
Comment (cjwatson) [inline test/routes/src/server/routes/t_launchpad.js]: This pair is unnecessarily wordy given that the containing `describe` already does almost the same thing.  I think you could just have:

```
beforeEach(() => {
  getMemcached().set(`url:${repositoryUrl}`, snapUrl);
});
```

... and omit `afterEach` entirely.

File: test/routes/src/server/routes/t_webhook.js [hunk 5]
```diff
@@ -30,20 +29,7 @@ describe('The WebHook API endpoint', () => {
     overrides.clear('LP_API_TOKEN_SECRET');
   });
 
-  beforeEach(() => {
```
Comment (bartaz) [inline test/routes/src/server/routes/t_webhook.js]: Webhook tests should not be concerned about memcached at all.

File: test/routes/src/server/routes/t_launchpad.js [hunk 6]
```diff
@@ -315,21 +315,6 @@ describe('The Launchpad API endpoint', () => {
   });
 
   describe('find snap route', () => {
-    beforeEach(() => {
-      const memcachedStub = { cache: {} };
```
Comment (bartaz) [inline test/routes/src/server/routes/t_launchpad.js]: Removed unnecessary memcached stubs.

File: test/routes/src/server/routes/t_launchpad.js [hunk 7]
```diff
@@ -473,25 +458,58 @@ describe('The Launchpad API endpoint', () => {
...
+        memcachedStub.get = (key, callback) => {
+          callback(undefined, memcachedStub.cache[key]);
+        };
+        memcachedStub.set = (key, value, lifetime, callback) => {
+          memcachedStub.cache[key] = value;
+          callback(undefined, true);
+        };
+        setMemcached(memcachedStub);
+
+        memcachedStub.cache[`url:${repositoryUrl}`] = snapUrl;
```
Comment (bartaz) [inline test/routes/src/server/routes/t_launchpad.js]: Testing if values from memcached are properly returned.

File: src/server/helpers/memcached.js (hunk 3 above)
Comment (cjwatson) [inline src/server/helpers/memcached.js]: I think it's OK to leave this up to the config.

**Pull request merged**